README.md
<!--
  Copyright (c) 2023 Eliah Kagan

  Permission to use, copy, modify, and/or distribute this software for any
  purpose with or without fee is hereby granted.

  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
  REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
  AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
  LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
  OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
  PERFORMANCE OF THIS SOFTWARE.
-->

# subaudit: Subscribe and unsubscribe for specific audit events

[Audit hooks](https://docs.python.org/3/library/audit_events.html) in Python
are called on all events, and they remain in place until the interpreter shuts
down.

This library provides a higher-level interface that allows listeners to be
subscribed to specific audit events, and unsubscribed from them. It provides
[context managers](#basic-usage) for using that interface with a convenient
notation that ensures the listener is unsubscribed. The context managers are
reentrant—you can nest `with`-statements that listen to events. By default, a
single audit hook is used for any number of events and listeners.

The primary use case for this library is in writing test code.

## License

subaudit is licensed under [0BSD](https://spdx.org/licenses/0BSD.html), which
is a [“public-domain
equivalent”](https://en.wikipedia.org/wiki/Public-domain-equivalent_license)
license. See
[**`LICENSE`**](https://github.com/EliahKagan/subaudit/blob/main/LICENSE).

## Compatibility

The subaudit library can be used to observe [audit events generated by the
Python interpreter and standard
library](https://docs.python.org/3/library/audit_events.html), as well as
custom audit events. It requires Python 3.7 or later. It is most useful on
Python 3.8 or later, because [audit events were introduced in Python
3.8](https://peps.python.org/pep-0578/). On Python 3.7, subaudit uses [the
*sysaudit* library](https://pypi.org/project/sysaudit/) to support audit
events, but the Python interpreter and standard library still do not provide
any events, so only custom events can be used on Python 3.7.

To avoid the performance cost of explicit locking in the audit hook, [some
operations are assumed atomic](#locking). I believe these assumptions are
correct for CPython, as well as PyPy and some other implementations, but there
may exist Python implementations on which these assumptions don’t hold.

## Installation  <!-- Maybe remove this section once badges are added. -->

Install [the `subaudit` package (PyPI)](https://pypi.org/project/subaudit/) in
your project’s environment.

## Basic usage

### The `subaudit.listening` context manager

The best way to use subaudit is usually the `listening` context manager.

```python
import subaudit

def listen_open(path, mode, flags):
    ...  # Handle the event.

with subaudit.listening('open', listen_open):
    ...  # Do something that may raise the event.
```

The listener—here, `listen_open`—is called with the event arguments each time
the event is raised. They are passed to the listener as separate positional
arguments (not as an `args` tuple).

In tests, it is convenient to use [`Mock`
objects](https://docs.python.org/3/library/unittest.mock.html#the-mock-class)
as listeners, because they record calls, provide a
[`mock_calls`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.mock_calls)
attribute to see the calls, and provide [various `assert_*`
methods](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.assert_called)
to make assertions about the calls:

```python
from unittest.mock import ANY, Mock
import subaudit

with subaudit.listening('open', Mock()) as listener:
    ...  # Do something that may raise the event.

listener.assert_any_call('/path/to/file.txt', 'r', ANY)
```

Note how, when the `listening` context manager is entered, it returns the
`listener` that was passed in, for convenience.

### The `subaudit.extracting` context manager

You may want to extract some information about calls to a list:

```python
from dataclasses import InitVar, dataclass
import subaudit

@dataclass(frozen=True)
class PathAndMode:  # Usually strings. See examples/notebooks/open_event.ipynb.
    path: str
    mode: str
    flags: InitVar = None  # Opt not to record this argument.

with subaudit.extracting('open', PathAndMode) as extracts:
    ...  # Do something that may raise the event.

assert PathAndMode('/path/to/file.txt', 'r') in extracts
```

The extractor—here, `PathAndMode`—can be any callable that accepts the event
args as separate positional arguments. Entering the context manager returns an
initially empty list, which will be populated with *extracts* gleaned from the
event args. Each time the event is raised, the extractor is called and the
object it returns is appended to the list.

### `subaudit.subscribe` and `subaudit.unsubscribe`

Although you should usually use the `listening` or `extracting` context
managers instead, you can subscribe and unsubscribe listeners without a context
manager:

```python
import subaudit

def listen_open(path, mode, flags):
    ...  # Handle the event.

subaudit.subscribe('open', listen_open)
try:
    ...  # Do something that may raise the event.
finally:
    subaudit.unsubscribe('open', listen_open)
```

Attempting to unsubscribe a listener that is not subscribed raises
`ValueError`. Currently, subaudit provides no feature to make this succeed
silently instead. But you can suppress the exception:

```python
with contextlib.suppress(ValueError):
    subaudit.unsubscribe('glob.glob', possibly_subscribed_listener)
```

## Nesting

To unsubscribe a listener from an event, it must be subscribed to the event.
Subject to this restriction, calls to [`subscribe` and
`unsubscribe`](#subauditsubscribe-and-subauditunsubscribe) can happen in any
order, and [`listening`](#the-subauditlistening-context-manager) and
[`extracting`](#the-subauditextracting-context-manager) may be arbitrarily
nested.

`listening` and `extracting` support reentrant use with both the same event and
different events. Here’s an example with three `listening` contexts:

```python
from unittest.mock import Mock, call

listen_to = Mock()  # Let us assert calls to child mocks in a specific order.

with subaudit.listening('open', print):  # Print all open events' arguments.
    with subaudit.listening('open', listen_to.open):  # Log opening.
        with subaudit.listening('glob.glob', listen_to.glob):  # Log globbing.
            ...  # Do something that may raise the events.

assert listen_to.mock_calls == ...  # Assert a specific order of calls.
```

(That is written out to make the nesting clear. You could also use a single
`with`-statement with commas.)

Here’s an example with both `listening` and `extracting` contexts:

```python
from unittest.mock import Mock, call

def extract(*args):
    return args

with (
    subaudit.extracting('pathlib.Path.glob', extract) as glob_extracts,
    subaudit.listening('pathlib.Path.glob', Mock()) as glob_listener,
    subaudit.extracting('pathlib.Path.rglob', extract) as rglob_extracts,
    subaudit.listening('pathlib.Path.rglob', Mock()) as rglob_listener,
):
    ...  # Do something that may raise the events.

# Assert something about, or otherwise use, the mocks glob_listener and
# rglob_listener, as well as the lists glob_extracts and rglob_extracts.
...
```

(That example uses [parenthesized context
managers](https://docs.python.org/3/whatsnew/3.10.html#parenthesized-context-managers),
which were introduced in Python 3.10.)

## Specialized usage

### `subaudit.Hook` objects

Each instance of the `subaudit.Hook` class represents a single audit hook that
supports subscribing and unsubscribing listeners for any number of events, with
methods corresponding to the four top-level functions listed above. Separate
`Hook` instances use separate audit hooks. The `Hook` class exists for three
purposes:

- It supplies the behavior of [the top-level `listening`, `extracting`,
  `subscribe`, and `unsubscribe` functions](#basic-usage), which correspond to
  the same-named methods on a global `Hook` instance.
- It allows multiple audit hooks to be used, for special cases where that might
  be desired.
- It facilitates customization, as detailed below.

The actual audit hook that a `Hook` object encapsulates is not installed until
the first listener is subscribed. This happens on the first call to its
`subscribe` method, or the first time one of its context managers (from calling
its `listening` or `extracting` method) is entered. This is also true of the
global `Hook` instance used by the top-level functions—merely importing
`subaudit` does not install an audit hook.

Whether the top-level functions are bound methods of a `Hook` instance, or
delegate in some other way to those methods on an instance, is currently
considered an implementation detail.

### Deriving from `Hook`

You can derive from `Hook` to provide custom behavior for subscribing and
unsubscribing, by overriding the `subscribe` and `unsubscribe` methods. You can
also override the `listening` and `extracting` methods, though that may be less
useful. Overridden `subscribe` and `unsubscribe` methods are automatically used
by `listening` and `extracting`.

Whether `extracting` uses `listening`, or directly calls `subscribe` and
`unsubscribe`, is currently considered an implementation detail.

### Locking

Consider two possible cases of race conditions:

#### 1. Between audit hook and `subscribe`/`unsubscribe` (audit hook does not lock)

In this scenario, a `Hook` object’s installed audit hook runs at the same time
as a listener is subscribed or unsubscribed.

This is likely to occur often and it cannot be prevented, because audit hooks
are called for all audit events. For the same reason, locking in the audit hook
has performance implications. Instead of having audit hooks take locks,
subaudit relies on each of these operations being atomic:

- *Writing an attribute reference, when it is a simple write to an instance
  dictionary or a slot.* Writing an attribute need not be atomic when, for
  example, `__setattr__` has been overridden.
- *Writing or deleting a ``str`` key in a dictionary whose keys are all of the
  built-in ``str`` type.* Note that the search need not be atomic, but the
  dictionary must always be observed to be in a valid state.

The audit hook is written, and the data structures it uses are selected, to
avoid relying on more than these assumptions.

#### 2. Between calls to `subscribe`/`unsubscribe` (by default, they lock)

In this scenario, two listeners are subscribed at a time, or unsubscribed at a
time, or one listener is subscribed while another (or the same) listener is
unsubscribed.

This is less likely to occur and much easier to avoid. But it is also harder to
make safe without a lock. Subscribing and unsubscribing are unlikely to happen
at a *sustained* high rate, so locking is unlikely to be a performance
bottleneck. So, *by default*, subscribing and unsubscribing are synchronized
with a
[`threading.Lock`](https://docs.python.org/3/library/threading.html#threading.Lock),
to ensure that shared state is not corrupted.

You should not usually change this. But if you want to, you can construct a
`Hook` object by calling `Hook(sub_lock_factory=...)` instead of `Hook`, where
`...` is a type, or other context manager factory, to be used instead of
`threading.Lock`. In particular, to disable locking, pass
[`contextlib.nullcontext`](https://docs.python.org/3/library/contextlib.html#contextlib.nullcontext).

## Functions related to compatibility

As [noted above](#compatibility), Python supports audit hooks [since
3.8](https://peps.python.org/pep-0578/). For Python 3.7, but not Python 3.8 or
later, the subaudit library declares
[sysaudit](https://pypi.org/project/sysaudit/) as a dependency.

### `subaudit.addaudithook` and `subaudit.audit`

subaudit exports `addaudithook` and `audit` functions.

- On Python 3.8 and later, they are
  [`sys.addaudithook`](https://docs.python.org/3/library/sys.html#sys.addaudithook)
  and [`sys.audit`](https://docs.python.org/3/library/sys.html#sys.audit).
- On Python 3.7, they are
  [`sysaudit.addaudithook`](https://sysaudit.readthedocs.io/en/latest/#sysaudit.addaudithook)
  and
  [`sysaudit.audit`](https://sysaudit.readthedocs.io/en/latest/#sysaudit.audit).

subaudit uses `subaudit.addaudithook` when it adds its own audit hook (or all
its own hooks, if you use additional [`Hook`](#subaudithook-objects) instances
besides the global one implicitly used by [the top-level
functions](#basic-usage)). subaudit does not itself use `subaudit.audit`, but
it is whichever `audit` function corresponds to `subaudit.addaudithook`.

### `@subaudit.skip_if_unavailable`

The primary use case for subaudit is in writing unit tests, to assert that
particular events have been raised or not raised. Usually these are [“built in”
events](https://docs.python.org/3.8/library/audit_events.html)—those raised by
the Python interpreter or standard library. But the sysaudit library doesn’t
backport those events, which would not really be feasible to do.

For this reason, tests that particular audit events did or didn’t occur—such as
a test that a file has been opened by listening to the `open` event—should
typically be skipped when running a test suite on Python 3.7.

**When using the [unittest](https://docs.python.org/3/library/unittest.html)
framework**, you can apply the `@skip_if_unavailable` decorator to a test class
or test method, so it is skipped prior to Python 3.8 with a message explaining
why. For example:

```python
import unittest
from unittest.mock import ANY, Mock
import subaudit

class TestSomeThings(unittest.TestCase):
    ...

    @subaudit.skip_if_unavailable  # Skip this test if < 3.8, with a message.
    def test_file_is_opened_for_read(self):
        with subaudit.listening('open', Mock()) as listener:
            ...  # Do something that may raise the event.

        listener.assert_any_call('/path/to/file.txt', 'r', ANY)

    ...

@subaudit.skip_if_unavailable  # Skip the whole class if < 3.8, with a message.
class TestSomeMoreThings(unittest.TestCase):
    ...
```

It could be useful also to have a conditional xfail ([expected
failure](https://docs.python.org/3/library/unittest.html#unittest.expectedFailure))
decorator for unittest—and, more so,
[marks](https://docs.pytest.org/en/7.1.x/how-to/mark.html) for
[pytest](https://docs.pytest.org/) providing specialized
[skip](https://docs.pytest.org/en/7.3.x/how-to/skipping.html#skipping-test-functions)/[skipif](https://docs.pytest.org/en/7.3.x/how-to/skipping.html#id1)
and
[xfail](https://docs.pytest.org/en/7.3.x/how-to/skipping.html#xfail-mark-test-functions-as-expected-to-fail)—but
subaudit does not currently provide them. Of course, in pytest, you can still
use the [`@pytest.mark.skip` and
`@pytest.mark.xfail`](https://docs.pytest.org/en/7.3.x/how-to/skipping.html)
decorators, by passing `sys.version_info < (3, 8)` as the condition.

## Overview by level of abstraction

From higher to lower level, from the perspective of the top-level
[`listening`](#the-subauditlistening-context-manager) and
[`extracting`](#the-subauditextracting-context-manager) functions:

- [`subaudit.extracting`](#the-subauditextracting-context-manager) - context
  manager that listens and extracts to a list
- [`subaudit.listening`](#the-subauditlistening-context-manager) - context
  manager to subscribe and unsubscribe a custom listener *(usually use this)*
- [`subaudit.subscribe` and
  `subaudit.unsubscribe`](#subauditsubscribe-and-subauditunsubscribe) -
  manually subscribe/unsubscribe a listener
- [`subaudit.Hook`](#subaudithook-objects) - abstraction around an audit hook
  allowing subscribing and unsubscribing for specific events, with
  `extracting`, `listening`, `subscribe`, and `unsubscribe` instance methods
- [`subaudit.addaudithook`](#subauditaddaudithook-and-subauditaudit) - trivial
  abstraction representing whether the function from `sys` or `sysaudit` is
  used
- [`sys.addaudithook`](https://docs.python.org/3/library/sys.html#sys.addaudithook)
  or
  [`sysaudit.addaudithook`](https://sysaudit.readthedocs.io/en/latest/#sysaudit.addaudithook)
  \- *not part of subaudit* - install a [PEP
  578](https://peps.python.org/pep-0578/) audit hook

This list is not exhaustive. For example,
[`@skip_if_unavailable`](#subauditskip_if_unavailable) is not part of that
conceptual hierarchy.

## Acknowledgements

I’d like to thank:

- [**Brett Langdon**](https://github.com/brettlangdon), who wrote the
  [sysaudit](https://github.com/brettlangdon/sysaudit) library (which subaudit
  [uses on 3.7](#compatibility)).

- [**David Vassallo**](https://github.com/dmvassallo), for reviewing pull
  requests about testing using audit hooks in [a project we have collaborated
  on](https://github.com/dmvassallo/EmbeddingScratchwork), which helped me to
  recognize what kinds of usage were more or less clear and that it could be
  good to have a library like subaudit; and for coauthoring a
  `@skip_if_unavailable` decorator that had been used there, which motivated
  the one here.

## About the name

This library is called “subaudit” because it provides a way to effectively
*sub*scribe to and un*sub*scribe from a *sub*set of audit events rather than
all of them.


CHANGELOG.md
<!-- SPDX-License-Identifier: 0BSD -->

# Changelog

*Major* changes are documented in this file.

## [0.1.0](https://github.com/EliahKagan/subaudit/releases/tag/0.1.0) - 2023-05-06

*Initial release.*

### Added

- `Hook` class.
- Top-level `subscribe`, `unsubscribe`, `listening`, and `extracting`
  functions.
- `@skip_if_unavailable` decorator.
- Full test suite.
- Detailed readme.
- A few example notebooks and scripts.


tests/_helpers.py
# Copyright (c) 2023 Eliah Kagan
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
# REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
# AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
# INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
# LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
# OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
# PERFORMANCE OF THIS SOFTWARE.

"""
Shared helpers, other than fixtures or their return types, for making fixtures.

The ``conftest`` module docstring explains what should go there and here.
"""

__all__ = ['ShortStrEnum']

import enum


class ShortStrEnum(enum.Enum):
    """An ``Enum`` whose instances' unqualified names are their ``str``s."""

    def __str__(self) -> str:
        """Unqualified name of the instance."""
        return self.name


examples/scripts/redirect.py
#!/usr/bin/env python

# SPDX-License-Identifier: 0BSD

"""True redirection."""

import os

import subaudit

_STDIN_FILENO = 0
"""File descriptor of real standard input."""


def main() -> None:
    """Run the redirection experiment."""
    with subaudit.extracting('builtins.input/result', lambda x: x) as extracts:
        old_stdin_fd = os.dup(_STDIN_FILENO)
        os.close(_STDIN_FILENO)
        try:
            with open('LICENSE', encoding='utf-8') as file:
                assert file.fileno() == _STDIN_FILENO
                result = input('Should read first line of file: ')
                print(f'Got: {result!r}')
        finally:
            os.dup2(old_stdin_fd, _STDIN_FILENO)
            os.close(old_stdin_fd)

        result = input('Should read from original stdin: ')
        print(f'Got: {result!r}')

    print(f'extracts={extracts!r}')  # f-strings in 3.7 don't support {var=}.


if __name__ == '__main__':
    main()


src/subaudit/__init__.py
# Copyright (c) 2023 Eliah Kagan
#
# Permission to use, copy, modify, and/or distribute this software for any
# purpose with or without fee is hereby granted.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
# REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
# AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
# INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
# LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
# OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
# PERFORMANCE OF THIS SOFTWARE.

"""
subaudit: Subscribe and unsubscribe for specific audit events.

This library provides a higher-level interface that allows listeners to be
subscribed to specific audit events, and unsubscribed from them. It provides
context managers for using that interface with a convenient notation that
ensures the listener is unsubscribed. The context managers are reentrant — you
can nest ``with``-statements that listen to events. By default, a single audit
hook is used for any number of events and listeners.

The primary use case for this library is in writing test code. Although it can
be used and customized in a few ways, the top-level ``subaudit.listening`` and
``subaudit.extracting`` context manager functions are usually sufficient.

Example usage:

    import subaudit

    def listen_open(path, mode, flags):
        ...  # Handle the event.

    with subaudit.listening('open', listen_open):
        ...  # Do something that may raise the event.

See the project ``README.md`` for more information on usage.
"""

__all__ = [
    'LockContextManagerFactory',
    'audit',
    'addaudithook',
    'Hook',
    'subscribe',
    'unsubscribe',
    'listening',
    'extracting',
    'skip_if_unavailable',
]

import contextlib
import sys
import threading
from typing import (
    Any,
    Callable,
    ContextManager,
    Generator,
    List,
    MutableMapping,
    NoReturn,
    Optional,
    Tuple,
    TypeVar,
    Union,
)
import unittest

if sys.version_info < (3, 8):
    from sysaudit import addaudithook, audit
else:
    from sys import addaudithook, audit

_F = TypeVar('_F', bound=Callable[..., None])
"""Type variable used to represent the callable type of an event listener."""

_R = TypeVar('_R')
"""Type variable used to represent the return type of an extractor."""

LockContextManager = Union[ContextManager[None], ContextManager[bool]]
"""
Type alias covering ``None``-returning and ``bool``-returning context managers.

``Lock``, ``RLock``, and other synchronization primitives' ``__enter__``
methods return ``bool`` (because ``acquire`` does). Most context managers that
may be used in their place, including ``contextlib.nullcontext``, return
``None``.

This alias differs subtly from ``Optional[bool]``. With ``Optional[bool]``, the
same object could return a ``bool`` on some calls and ``None`` on others.
"""

LockContextManagerFactory = Callable[[], LockContextManager]
"""
Type alias for classes or factory functions returning a ``LockContextManager``.
"""


class Hook:
    """
    Audit hook wrapper. Subscribes and unsubscribes specific-event listeners.

    Listeners are subscribed and unsubscribed for specific auditing events.
    Only one audit hook (per ``Hook`` instance) is used. It is installed the
    first time a listener is subscribed to any event via the ``Hook`` instance;
    if the instance is never used, no audit hook is installed. A program rarely
    needs multiple ``Hook`` instances, even with many listeners and events. The
    ``subaudit`` module provides top-level ``subscribe``, ``unsubscribe``,
    ``listening``, and ``extracting`` functions, which use a pre-created
    ``Hook`` instance.

    The ``subscribe`` and ``unsubscribe`` methods, but not the installed audit
    hook, are by default protected by a mutex. The audit hook can be called at
    any time, including as ``subscribe`` or ``unsubscribe`` runs: it is called
    on all audit events, filtering for those of interest. However, if the
    Python interpreter is CPython — or another implementation where writing an
    attribute reference is atomic, and writing or deleting an item in a dict
    with string keys is atomic — then the ``Hook``'s state should not be
    corrupted. In short, on CPython, strange behavior and segfaults shouldn't
    happen due to an event firing, even if a listener subscribes or
    unsubscribes at the same time.

    ``Hook`` objects are not optimized for the case of an event having a large
    number of listeners. This is because a ``Hook`` stores each event's
    listeners in an immutable sequence, rebuilt each time a listener is
    subscribed or unsubscribed. (This is part of how consistent state is
    maintained, so the audit hook doesn't need to synchronize with subscribe
    and unsubscribe.) Subscribing N listeners to the same event without
    unsubscribing takes O(N²) time. If you need more than a couple hundred
    listeners on the same event at the same time, especially if you also
    frequently subscribe and unsubscribe listeners to that same event, this may
    be the wrong tool.
    """

    __slots__ = ('_lock', '_hook_installed', '_table')

    _lock: LockContextManager
    """
    Mutex or other context manager. Protects ``subscribe`` and ``unsubscribe``.
    """

    _hook_installed: bool
    """Whether the audit hook is installed yet."""

    _table: MutableMapping[str, Tuple[Callable[..., None], ...]]
    """Table that maps each event to its listeners."""

    def __init__(
        self, *, sub_lock_factory: Optional[LockContextManagerFactory] = None,
    ) -> None:
        """
        Make an audit hook wrapper, which will use its own audit hook.

        :param sub_lock_factory: If passed, this is called instead of
            ``threading.Lock`` to create a context manager object, which is
            used as a mutex during subscribing and unsubscribing. To forgo
            locking, pass ``contextlib.nullcontext``.
        """
        if sub_lock_factory is None:
            sub_lock_factory = threading.Lock
        self._lock = sub_lock_factory()
        self._hook_installed = False
        self._table = {}

    def __repr__(self) -> str:
        """Representation for debugging. Not runnable as Python code."""
        return f'<{type(self).__name__} at {id(self):#x}: {self._summarize()}>'

    def subscribe(self, event: str, listener: Callable[..., None]) -> None:
        """
        Attach a detachable listener to an event.

        :param event: The name of the event to subscribe the listener to.
        :param listener: The callable listener to be subscribed to the event.

        Overriding ``subscribe`` and ``unsubscribe`` will customize the
        behavior of ``listening`` and ``extracting``.
        """
        # Produce a suitable key for the event, raising TypeError if we can't.
        if type(event) is not str:
            if not isinstance(event, str):
                raise TypeError(f"event must be 'str', not {type(event)!r}")
            event = f'{event!s}'  # Exactly str, even when str(event) isn't.

        with self._lock:
            if not self._hook_installed:
                addaudithook(self._hook)
                self._hook_installed = True

            old_listeners = self._table.get(event, ())
            self._table[event] = (*old_listeners, listener)

    def unsubscribe(self, event: str, listener: Callable[..., None]) -> None:
        """
        Detach a listener that was attached to an event.

        :param event: The name of the event to unsubscribe the listener from.
        :param listener: The callable listener to be unsubscribed from the
            event.

        Overriding ``subscribe`` and ``unsubscribe`` will customize the
        behavior of ``listening`` and ``extracting``.
        """
        with self._lock:
            try:
                listeners = self._table[event]
            except KeyError:
                self._fail_unsubscribe(event, listener)

            # Search in reverse, to find the most recent matching listener.
            try:
                reverse_last_index = listeners[::-1].index(listener)
            except ValueError:
                self._fail_unsubscribe(event, listener)

            # Remove it. (Use positive indexing to avoid slicing from "-0".)
            last_index = len(listeners) - 1 - reverse_last_index
            listeners = listeners[:last_index] + listeners[(last_index + 1):]

            # Replace the row, unless empty, in which case remove it.
            if listeners:
                self._table[event] = listeners
            else:
                del self._table[event]

    def listening(self, event: str, listener: _F) -> ContextManager[_F]:
        """
        Context manager to subscribe and unsubscribe an event listener.

        :param event: The name of the event to subscribe the listener to.
        :param listener: The callable listener to be subscribed to the event.
        :return: A context manager that subscribes the listener when entered
            and unsubscribes the listener when exited.
        """
        return self._make_listening(event, listener)

    def extracting(
        self, event: str, extractor: Callable[..., _R],
    ) -> ContextManager[List[_R]]:
        """
        Context manager to provide a list of custom-extracted event data.

        This function returns a context manager object. Entering the context
        manager returns an initially empty list. The list is populated with
        extracts produced by calling ``extractor`` with the event args. Each
        time ``event`` occurs, an extract is appended to the list, until the
        context manager is exited.

        :param event: The name of the event to extract data from.
        :param extractor: The callable extractor that selects data from event
            args.
        :return: A context manager that starts extracting data from the event
            when entered, and stops when exited.
        """
        return self._make_extracting(event, extractor)

    def _hook(self, event: str, args: Tuple[Any, ...]) -> None:
        """Single audit hook used for all events and handlers."""
        try:
            # Subscripting a dict with str keys should be sufficiently
            # protected by the GIL in CPython. This doesn't protect the table
            # rows. But those are tuples that we always replace, rather than
            # lists that we mutate, so we should observe consistent state.
            listeners = self._table[event]
        except KeyError:
            return

        for listener in listeners:
            listener(*args)

    def _summarize(self) -> str:
        """
        Summarize the state of the ``Hook`` instance. Used as part of the repr.

        For now, just include info CPython lets us get safely without a lock.
        """
        if not self._hook_installed:
            return 'audit hook not installed'

        num_events = len(self._table)
        if num_events == 1:
            return f'watching {num_events} event'
        return f'watching {num_events} events'

    @staticmethod
    def _fail_unsubscribe(
        event: str, listener: Callable[..., None],
    ) -> NoReturn:
        """Raise an error for an unsuccessful attempt to detach a listener."""
        raise ValueError(f'{event!r} listener {listener!r} never subscribed')

    @contextlib.contextmanager
    def _make_listening(
        self, event: str, listener: _F,
    ) -> Generator[_F, None, None]:
        """
        Helper for ``listening``.

        Callers shouldn't assume ``listening`` returns
        ``GeneratorContextManager``. This helper allow ``listening`` to have
        the desired type annotations. Subclasses may override ``listening`` but
        shouldn't override or call this.
        """
        self.subscribe(event, listener)
        try:
            yield listener
        finally:
            self.unsubscribe(event, listener)

    @contextlib.contextmanager
    def _make_extracting(
        self, event: str, extractor: Callable[..., _R],
    ) -> Generator[List[_R], None, None]:
        """
        Helper for ``extracting``.

        Callers shouldn't assume ``extracting`` returns
        ``GeneratorContextManager``. This helper allows ``extracting`` to have
        the desired type annotations. Subclasses may override ``extracting``
        but shouldn't override or call this.
        """
        extracts: List[_R] = []

        def append_extract(*args: Any) -> None:
            extracts.append(extractor(*args))

        with self.listening(event, append_extract):
            yield extracts


_global_instance = Hook()
"""
``Hook`` instance used by the top-level functions.

The module-level ``subscribe``, ``unsubscribe``, ``listening``, and
``extracting`` functions use this instance. This should not be confused with
the behavior of each ``Hook`` object in installing (at most) one actual
auditing event hook.
"""

subscribe = _global_instance.subscribe
unsubscribe = _global_instance.unsubscribe
listening = _global_instance.listening
extracting = _global_instance.extracting


skip_if_unavailable = unittest.skipIf(
    sys.version_info < (3, 8),
    'Python Runtime Audit Hooks (PEP 578) were introduced in Python 3.8.',
)
"""
Skip a ``unittest`` test if the standard library doesn't have audit events.
"""
